Use diff not log for multi commit diffing

tidy-dev opened the pull request:
## Description
This is first step in the direction to support multi commit diffing for that we need to use `git diff`. This PR is to setup to use `git diff` as opposed to `git log` when more than one commit is selected. This works for one commit as well but due to this being a very impactful change if there are regressions as diff generation is core to the app, I am leaving `git log` in place for single commits. We may decide to remove it once confidence in this approach is built through implementing multi commit diffing. 

Scenarios considered/tested for a single commit: 
- adding
- changing
- renaming
- images
- initial commit
- squashing onto initial commit

## Release notes
Notes: no-notes

tidy-dev committed:
Add feature flag for multi commit diff work

tidy-dev committed:
Add method get diff between two commits

tidy-dev commented on app/src/lib/git/diff.ts:
```diff
@@ -133,6 +133,59 @@ export async function getCommitDiff(
+    commitsToUse.length === 1
+      ? `${commitsToUse.at(0)}^..${commitsToUse.at(0)}`
+      : `${commitsToUse.at(-1)}..${commitsToUse.at(0)}`
+  const oldestCommit = `${commitsToUse.at(-1)}`
```
This bit is the major downside to `git diff` over `git log`. Thoughts? 
1) Could leave as is and use this constant.. some googling says it has always been not likely to change.
2) Use `git hash-object -t tree /dev/null` (save the result after first call).. a little future proofing in case this were to ever change.

niik commented on app/src/lib/git/diff.ts:
```diff
@@ -133,6 +133,59 @@ export async function getCommitDiff(
+  }
+
+  // hacky? We could derive - git hash-object -t tree /dev/null
+  const emptyTreeHash = '4b825dc642cb6eb9a060e54bf8d69288fbee4904'
```
Using the null tree is fine, in fact we've already defined this as a constant in `diff-index.ts` (NullTreeSHA). Might as well pull it out of there and into this file and export it along with a little blurb of documentation.

niik commented on app/src/ui/dispatcher/dispatcher.ts:
```diff
@@ -3552,7 +3558,16 @@ export class Dispatcher {
+              ? (await getCountCommitsInBranch(
+                  repository,
+                  status.currentBranch
+                )) === 1
```
Counting commits in a repository can be very expensive for large repositories. Instead of using `rev-list --count` here we should look at the commit in question and see if it has any parents. If we have to know whether this commit is the first commit here a specialized method to run `git log -1 --format=%P SHA --` would be my preference

tidy-dev replied:
:D One of the reasons I split this off into its own PR so early is I wanted to get your all feedback about stuff like this! That makes total sense

tidy-dev replied:
Moot since don't need to track this state

niik commented on app/src/ui/history/commit-list.tsx:
```diff
@@ -272,16 +275,26 @@ export class CommitList extends React.Component<ICommitListProps, {}> {
+      this.props.commitSHAs.findIndex(sha => sha === v.sha)
+    )
+    const selectionContainsInitialCommit =
+      this.getLastRetainedCommitRef(indexes) === null
```
Is this really doing what we expect it to? Looking at the `getLastRetainedCommitRef` it seems to look at the last commit in the `commitSHAs` list is among the selected ones. That doesn't _necessary_ mean that the first commit of the branch is selected since `commitSHAs` is loaded incrementally.

So if you go to the history tab and hit <kbd>Cmd</kbd>+<kbd>A</kbd> to select all commits the `selectionContainsInitialCommit` will be true and we will diff against the null tree without having selected the initial commit.

I think what we'd want to do here is something like

```suggestion
    const selectionContainsInitialCommit =
      selectedCommits.at(-1)?.parentSHAs.length === 0

tidy-dev replied:
Rendered moot since changed approach to try, check for bad revision, and then retry with null tree sha.

niik reviewed (changes requested):
This is a tricky one @tidy-dev but I'm wondering if, given how rare it is to diff the initial commit, we should optimize for the case where the initial commit is not part of the selection and get rid of the `selectionContainsInitialCommit` state keeping entirely.

Instead we could attempt to run the diff and detect if it fails due to an invalid revision and in that case diff against the null tree instead.

> ⚠️ Warning. This is a very impactful change if there are regressions as diff generation is core to the app. 😟 Thus, open for any suggestions for different scenarios to test. (Also, is feature flagged)

We could keep using the `log` approach for single-commit diffs for the foreseeable future and only switch to the diff based approach once we've gained enough confidence in it. This has the added benefit of us not having to solve the problem with `selectionContainsInitialCommit` at all for the time being. The more I think about it the more I like this approach, we avoid a lot of edge cases and get two specialized methods, one for single commit diffs, and one for commit ranges (endpoints really).

tidy-dev committed:
Use git (exec) instead of spawn

tidy-dev committed:
Use new commit diff method (feature flagged)

tidy-dev committed:
Retry diff will null tree sha for bad revisions

Update diff.ts

tidy-dev committed:
Use tried and true log for single commits

tidy-dev commented:
After some discussion, I have updated this PR to not keep track of a state variable to determine if the range is over an initial commit. Switched to @niik suggested approach of trying it and if it fails with a bad revision to retry it with the null tree sha. 

Additionally, added a clause to only use this new diff code if there is more than one sha being diffed (which at this point is not programmatically possible). **Thus, just pulling this PR to test is not enough to run this logic, you must remove the shas.length > 1 ( add in https://github.com/desktop/desktop/pull/14668/commits/772cbfc80d3f45cfaaed5988c0c7cc30231c3ccd) clause after the feature flag check, if you want to try using it to see diffs for a single commit**

tidy-dev commented on app/src/lib/git/diff.ts:
```diff
@@ -133,6 +136,75 @@ export async function getCommitDiff(
+  // initial commit of a branch) and therefore `SHA^` is not a valid reference.
+  // In which case, we will retry with the null tree sha.
+  if (result.gitError === GitError.BadRevision && useNullTreeSHA === false) {
+    const useNullTreeSHA = true
```
@niik We discussed adding another git call here to verify that the bad revision was caused by the oldest sha not having a parent and doing a non fatal error to track and make sure this isn't a thing, then removing the check in a month or so if we don't get any reports of it. 

I am leaning to this not being worth the effort since it "shouldn't" happen as Desktop is getting the sha's from the `git log` call and therefore they should exist and we shouldn't be get bad revisions (except in this case)

For anyone else following along, the negative consequence of not doing this would be that we could unintentionally attempt a diff for a commit range using the null tree sha that would be very large if the bad revision was due to the oldest commit reference being corrupted. (imagine going from desktop's latest commit in development to the initial commit :D)

niik replied:
That all sounds great. One comment on the line itself though, this is shadowing the original parameter by defining a new const over it. I'd rather we didn't do that as it's easily overlooked and can make it difficult to debug down the line once this is refactored enough. My preference would be to just pass `true` as the last argument to the `getCommitRangeDiff` call below. The comment above is clear enough about what's happening that we don't need the variable.

tidy-dev committed:
What is this error..

tidy-dev committed:
That error will already be handled by the `git` call

tidy-dev committed:
Had my earliest and latest commits inverted..

niik commented on app/src/lib/git/diff.ts:
```diff
@@ -27,6 +27,9 @@ import { getOldPathOrDefault } from '../get-old-path'
 import { getCaptures } from '../helpers/regex'
 import { readFile } from 'fs/promises'
 import { forceUnwrap } from '../fatal-error'
+import { git } from '.'
```
Let's import it straight from the file rather than the index to stay consistent with other imports. The index file is poorly maintained and I wonder if we shouldn't just get rid of it tbh.

```suggestion
import { git } from './core'
```

tidy-dev committed:
No shadowing

tidy-dev committed:
Import from core
